syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+{
+	sysptr_t ptr;
+	const syspage_prog_t *prog;
+
+	printf("\n\tSyspage:\n");
+	printf("\tImage size: 0x%08x\n", sysgen_common.syspage->hs.imgsz);
+	printf("\tSyspage size: 0x%08x\n", sysgen_common.syspage->size);
+	printf("\tKernel physical address: 0x%08x\n", sysgen_common.syspage->pkernel);
+	printf("\tConsole: 0x%02x\n", sysgen_common.syspage->console);
+	printf("\tPrograms:\n");
+	if (sysgen_common.syspage->progs != 0) {
+		ptr = sysgen_common.syspage->progs;
+		do {
+			prog = CAST_SYSPTR(syspage_prog_t *, ptr);
+			printf("\t\t%s\n", CAST_SYSPTR(char *, prog->argv));
+		} while ((ptr = prog->next) != sysgen_common.syspage->progs);
+	}
+	else {
+		printf("\t\tnot defined\n");
+	}
+}
+
+
+static void sysgen_help(const char *prog)
+{
+	printf("Usage: %s [options] to create binary syspage file or add syspage to image\n", prog);
+	printf("\t-s <pimg> <offs> <sz>   - syspage properties\n");
+	printf("\t    pimg - begining physical address of the target image\n");
+	printf("\t    offs - syspage's offset in the target image\n");
+	printf("\t    sz   - max syspage's size\n");
+	printf("\t-p <path>               - path to preinit script\n");
+	printf("\t-u <path>               - path to user script\n");
+	printf("\t-b <path>               - path to output syspage binary file\n");
+	printf("\t-i <path>               - path to image \n");
+	printf("\t-h                      - print help message\n");
+}
+
+
+int main(int argc, char *argv[])
+{
+	int opt, err;
+	char *endptr;
+	const char *preinitScript, *userScript, *binName, *imgName;
+
+	preinitScript = userScript = binName = imgName = NULL;
+
+	if (argc <= 1) {
+		sysgen_help(argv[0]);
+		return EXIT_FAILURE;
+	}
+
+	while ((opt = getopt(argc, argv, "s:p:u:b:i:h")) != -1) {
+		switch (opt) {
+			case 's':
+				if ((optind + 1) < argc) {
+					sysgen_common.pkernel = strtoul(argv[optind - 1], &endptr, 0);
+					if ((sysgen_common.pkernel == 0 && strlen(endptr) != 0) || (sysgen_common.pkernel != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong physical image address %s\n", argv[optind - 1]);
+						return -1;
+					}
+
+					sysgen_common.offs = strtoul(argv[optind], &endptr, 0);
+					if ((sysgen_common.offs == 0 && strlen(endptr) != 0) || (sysgen_common.offs != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong syspage offset %s\n", argv[optind]);
+						return -1;
+					}
+
+					sysgen_common.maxsz = strtoul(argv[optind + 1], &endptr, 0);
+					if ((sysgen_common.maxsz == 0 && strlen(endptr) != 0) || (sysgen_common.maxsz != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong syspage size %s\n", argv[optind + 1]);
+						return -1;
+					}
+				}
+				break;
+
+			case 'p':
+				preinitScript = optarg;
+				break;
+
+			case 'u':
+				userScript = optarg;
+				break;
+
+			case 'b':
+				binName = optarg;
+				break;
+
+			case 'i':
+				imgName = optarg;
+				break;
+
+			case 'h':
+			default:
+				sysgen_help(argv[0]);
+				return EXIT_SUCCESS;
```

Comment:
for `default` You should exit with failure, for `-h` with success

Reply:
done